spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -155,6 +158,20 @@ public long count(Class<?> entityClass) {
 		return findById(id, entityClass) != null;
 	}
 
+	@Override
+	public <A> A performTransaction(Function<DatastoreOperations, A> operations) {
+		if (!(this.datastore instanceof Datastore)) {
```

Comment:
Why did you make the change from `Datastore` to `DatastoreReaderWriter`? Otherwise you wouldn't need this check.

Reply:
The point is that the single-transaction-only DatastoreTemplate would only have a DatastoreReadWriter, and this check is meant to disallow transactions within transactions, which isn't allowed by Cloud Datastore.

Reply:
There are a number of methods in the `Datastore` interface that we may need in the future for `DatastoreTemplate`. Switching to a smaller set of operations available in `DatastoreReaderWriter` might be limiting.

To do this properly without casts, we may need to reflect the type hierarchy of the Datastore library. In other words, we might need a `DatatastoreOperations`, and a `TransactionOperations` interfaces, both inheriting from `DatastoreReadWriteOperations`.

However, if we're confident that we will never need anything that `Datastore` interface adds on top of `DatastoreReaderWriter`, then it's probably not worth the trouble.